### PR TITLE
Ding/add bidding

### PR DESCRIPTION
**Pull request by dingsheng-ong**
added bogo bidder (randomly buy/sell random amount of electricity at random price)

### CODE REVIEW

File: src/bidding/bogo_bidder.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    df = get_output_template()
+    for i in range(len(df)):
+        df.loc[i, "volume"] = (
+            random.random() * 100 + 50
+        )  # random values between 50 - 150
+        df.loc[i, "price"] = (
+            random.random() * 100 + 50
+        )  # random values between 50 - 150
+        df.loc[i, "type"] = random.choice(["BUY", "SELL"])
+    return df
```
Comment (fjebaker) [inline src/bidding/bogo_bidder.py:39]: I love it

File: server/routes/bidding.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        if "default" in keys:
+            keys.remove("default")
+        default = "none"
+        for key in keys:
+            if BIDDERS[key] == BIDDERS.get("default"):
+                default = key
+        mssg = {"bidders": keys, "default": default}
+        return mssg, 200
+    except Exception as e:
+        return {"message": str(e)}, 500
```
Comment (fjebaker) [inline server/routes/bidding.py:21]: Ah so the endpoint lets you choose which bidding algorithm to use?

**Comment by fjebaker** [comment]
Alright this is pretty cool, I like the idea of having a decorator for registering bidders, that's cooler than my idea of using metaclasses!

I see the database access in `src/bidders/util.py` is handled by the flask `g` object. Ideally, we would seperate talking to the database to the Flask app entirely, so that the scripts in `src` can be standalone and have any arbitrary data passed in. This  would let us switch out how they are used more easily, e.g. CLI, FAAS, via NodeRed exec, or as we have now, with Flask.

We can amend this in a later PR, I just thought I'd mention it here so there's a record.

File: server/routes/bidding.py (hunk 2 above)
Comment (dingsheng-ong) [reply server/routes/bidding.py:21]: that's my plan, but we can definitely change it in the future

**Pull request merged**